### PR TITLE
fix: parse attachments from streaming request

### LINKED ISSUE #3887: Attachments is undefined when using Direct Line App Service Extensions
## Versions

`4.14.1`.

Tested with Web Chat `4.14.0` (latest production as of now).

## Describe the bug

When connecting via DL ASE to a bot, uploading a file to the bot, should appear in the `activity.attachments` array. Instead, it is `undefined`.

## To Reproduce

1. Provision a new JS-based Echo Bot
2. Update the bot to support DL ASE
   - https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-node-bot?view=azure-bot-service-4.0
3. Make sure the bot is using `4.14.1` (or later) version of SDK
   - `npm install botbuilder@4.14.1`
4. Modify the `bot.js` as follow

```diff
-     const replyText = `Echo: ${context.activity.text || ''}`;
+     const replyText = `Echo: ${context.activity.text || ''}\n\n(With ${
+       (context.activity.attachments || []).length
+     } attachments, \`${typeof context.activity.attachments}\`)`;
```

5. Connect to the bot via DL ASE using Web Chat
   - Ping both `/.bot/` endpoint and `/api/messages` endpoint to make sure the bot is up and ready
   - HTML template at https://github.com/microsoft/BotFramework-WebChat/tree/master/samples/01.getting-started/i.protocol-direct-line-app-service-extension
   - Or, use the Web Chat Loader, see how in Additional Context
6. Upload a file to the bot

## Expected behavior

Bot should say, "Echo: (With 1 attachments, object)".

Instead, it said, "Echo: (With 0 attachments, undefined)".

### Direct Line App Service Extension

It says 0 attachments when connecting to the bot via DL ASE.

image

### Direct Line (Web Socket)

It works when connecting to the bot via Web Socket and uploading a file to it.

image

### Direct Line App Service Extension to a C# bot

It works when connecting to a C# bot via DL ASE.

Steps:

1. Provision a new C# bot with Echo Bot template
2. Update the bot to support DL ASE
   - https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-net-bot?view=azure-bot-service-4.0
3. Update `EchoBot.csproj` to use latest version of Bot SDK 4.14.1, the bot template was using 4.11.0

```diff
-   <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.11.0" />
+   <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.14.1" />
```

4. Update `EchoBot.cs`
   - You can use Project Kudu (a.k.a. SCM) to modify the code. Run `deploy.cmd` after your change.

```diff
-           var replyText = $"Echo: {turnContext.Activity.Text}";
+           var replyText = $"Echo: {turnContext.Activity.Text}\n\n(With {turnContext.Activity.Attachments?.Count} attachments)";
```

5. Connect to the bot via DL ASE using Web Chat
   - Ping both `/.bot/` endpoint and `/api/messages` endpoint to make sure the bot is up and ready
   - HTML template at https://github.com/microsoft/BotFramework-WebChat/tree/master/samples/01.getting-started/i.protocol-direct-line-app-service-extension
   - Or, use the Web Chat Loader, see how in Additional Context
6. Upload a file to the bot

image

### Test matrix

We tested the scenario through JS and C#, and via Web Socket or ASE.

As it only failed on JS via DL ASE, we believe this is an issue on Bot SDK.

| Protocol | JavaScript bot | C# bot |
| - | - | - |
| Direct Line Web Socket | ✔️ | ✔️ |
| Direct Line App Service Extension | ❌ | ✔️ |

### HTTP request to obtain a token for DL ASE

Paste the following into Fiddler Composer, replace `BOT_URL` and `BOT_SECRET`. The `BOT_SECRET` should be the Direct Line secret, not the DL ASE secret.

```http
POST http://BOT_URL.azurewebsites.net/.bot/v3/directline/tokens/generate HTTP/1.1
User-Agent: Fiddler
authorization: Bearer BOT_SECRET
Host: BOT_URL.azurewebsites.net

{"User":{"Id":"dl_1234"}}
```

### Test in Web Chat

To quickly test in Web Chat, you can use [Web Chat Loader](https://compulim.github.io/webchat-loader). Follow all the highlighted fields below. `BOT_TOKEN` should be the token you obtained from `/.bot/v3/directline/` endpoint (DL ASE specific token). You can use the HTTP request trace above to obtain a token.

image

### PR DESCRIPTION
**Pull request by joshgummersall**
Fixes #3887

### CODE REVIEW

File: libraries/botbuilder/src/botFrameworkAdapter.ts [hunk 1]
```diff
@@ -2011,8 +2012,15 @@ export class BotFrameworkAdapter
     }
 
     private async readRequestBodyAsString(request: IReceiveRequest): Promise<Activity> {
-        const contentStream = request.streams[0];
-        return await contentStream.readAsJson<Activity>();
+        const [activityStream, ...attachmentStreams] = request.streams;
+
+        const activity = await activityStream.readAsJson<Activity>();
+
+        activity.attachments = await Promise.all(
```
Comment (carlosscastro) [inline libraries/botbuilder/src/botFrameworkAdapter.ts:2019]: You probably are aware of this, but adding this note just in case -- attachments have a content type, and depending on that content type they could be json, cards, audio, text, etc. Another dimension is whether the attachment is inlined or has a url. I'd suggest to try this out with a few different contents and mechanism combinations to make sure we are covering the scenario families.

Comment (carlosscastro) [reply libraries/botbuilder/src/botFrameworkAdapter.ts:2019]: If you know for a fact that this code is independent of any corner case combination then that's fine. In C# there are 2 families of attachments to test to cover pretty much all cases, I'm just not sure what the minimum test base is for this code without spending a bit more time on it.

Comment (compulim) [reply libraries/botbuilder/src/botFrameworkAdapter.ts:2019]: I am not expert on this topic, you can refer to how C# SDK works.

FYI, someone told me, stream could be never-ending, e.g. a speech stream. That means `await Promise.all()` could (almost) never resolve.

Maybe we didn't go as far as this yet even in C#.

Comment (joshgummersall) [reply libraries/botbuilder/src/botFrameworkAdapter.ts:2019]: @compulim, the `IReceiveRequest`, in this context, is a discrete set of bytes from a socket. The socket, itself, is a stream that is never-ending.

@carlosscastro working on verifying now, thanks for the pointers! I think I have found another hiccup when adding `contentType`-aware stream reading.

Comment (carlosscastro) [reply libraries/botbuilder/src/botFrameworkAdapter.ts:2019]: @compulim good point! the endless stream could exist in the highest protocol layer, which assembles streams. However, this is a requirement that was part of the original design but it's not completely implemented, and we don't really use it for any practical scenario and many parts of the code don't really support that.

**Comment by joshgummersall** [comment]
@carlosscastro I've fixed the content type issue in the latest commit, please have a look and see if you agree with the changes. I verified that attachments coming in via DL ASE now appear well-formed (i.e. I'm getting a content type and either a JSON-parsed object for application/json or a string of chars otherwise).

File: libraries/botframework-streaming/src/assemblers/payloadAssembler.ts [hunk 2]
```diff
@@ -126,29 +126,41 @@ export class PayloadAssembler {
...
+        responsePayload: IResponsePayload | IRequestPayload,
+        receiveResponse: IReceiveResponse | IReceiveRequest
+    ): Promise<void> {
+        responsePayload.streams?.forEach((responseStream) => {
+            // There was a bug in how streams were handled. In .NET, the StreamDescription definiton mapped the
+            // "type" JSON property to `contentType`, but in Typescript there was no mapping. This is a workaround
+            // to resolve this inconsistency.
+            //
+            // .NET code:
+            // https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Streaming/Payloads/Models/StreamDescription.cs#L28-L29
```
Comment (stevengum) [inline libraries/botframework-streaming/src/assemblers/payloadAssembler.ts:152]: Can you use a permalink instead of referencing `main`? Thanks!

File: libraries/botbuilder/src/cloudAdapter.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 import * as z from 'zod';
 import type { BotFrameworkHttpAdapter } from './botFrameworkHttpAdapter';
-import { Activity, CloudAdapterBase, InvokeResponse, StatusCodes, TurnContext } from 'botbuilder-core';
+import { Activity, Attachment, CloudAdapterBase, InvokeResponse, StatusCodes, TurnContext } from 'botbuilder-core';
```
Comment (stevengum) [inline libraries/botbuilder/src/cloudAdapter.ts:6]: This import doesn't look to be necessary, going off of the changes in the file? Maybe it was accidentally left in? 😄

File: libraries/botbuilder/src/botFrameworkAdapter.ts [hunk 4]
```diff
@@ -37,6 +37,7 @@ import {
     StatusCodes,
     TokenResponse,
     TurnContext,
+    Attachment,
```
Comment (stevengum) [inline libraries/botbuilder/src/botFrameworkAdapter.ts:40]: Same feedback for this import 😄

**Review by stevengum — APPROVED** [review]
Minor nits around unused imports, otherwise it looks good!
